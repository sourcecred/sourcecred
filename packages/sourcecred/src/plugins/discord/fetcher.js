// @flow

import fetch from "isomorphic-fetch";
import * as Model from "./models";

export interface DiscordApi {
  guilds(): Promise<$ReadOnlyArray<Model.Guild>>;
  emojis(guild: Model.Snowflake): Promise<$ReadOnlyArray<Model.Emoji>>;
  channels(guild: Model.Snowflake): Promise<$ReadOnlyArray<Model.Channel>>;
  roles(guild: Model.Snowflake): Promise<$ReadOnlyArray<Model.Role>>;
  members(guild: Model.Snowflake): Promise<$ReadOnlyArray<Model.GuildMember>>;
  messages(
    channel: Model.Snowflake,
    after: Model.Snowflake,
    limit: number
  ): Promise<$ReadOnlyArray<Model.Message>>;
  reactions(
    channel: Model.Snowflake,
    message: Model.Snowflake,
    emoji: Model.Emoji
  ): Promise<$ReadOnlyArray<Model.Reaction>>;
}

const fetcherDefaults: FetcherOptions = {
  apiUrl: "https://discordapp.com/api",
  token: null,
  fetch,
};

type FetcherOptions = {|
  +apiUrl: string,
  +fetch: typeof fetch,
  +token: ?Model.BotToken,
|};

export class Fetcher implements DiscordApi {
  +_options: FetcherOptions;
  _timeout: number;

  constructor(opts?: $Shape<FetcherOptions>) {
    this._options = {...fetcherDefaults, ...opts};
    if (!this._options.token) {
      throw new Error("A BotToken is required");
    }
    this._timeout = 0;
  }

  _fetch(endpoint: string): Promise<Response> {
    const {apiUrl, token} = this._options;
    if (!token) {
      throw new Error("A BotToken is required");
    }
    const requestOptions = {
      method: "GET",
      headers: {
        Accept: "application/json",
        Authorization: `Bot ${token}`,
      },
    };
    const url = new URL(`${apiUrl}${endpoint}`).href;
    return this._options.fetch(url, requestOptions);
  }

  async _wait() {
    const currentTime = Date.now();
    // if timeout hasn't passed, we wait until it does
    if (currentTime < this._timeout) {
      const restartDate = new Date(this._timeout);
      const waitTime = this._timeout - currentTime;
      console.warn(
        `Discord Rate limit reached. Waiting for ${
          waitTime / 1000
        } seconds (until ${restartDate.toLocaleString()})`
      );
      await new Promise((resolve) => setTimeout(resolve, waitTime));
    }
  }

  _checkRateLimit(res: Response) {
    // The discord API returns this header to indicate how many requests remain
    // before hitting a cooldown.
    const rateLimitRemaining = Number(res.headers.get("x-ratelimit-remaining"));

    // The discord API returns this epoc time (in seconds) to let us know
    // when we can continue.
    const rateLimitReset = Number(res.headers.get("x-ratelimit-reset")) * 1000;

    // wait until the timeout passes  before attempting another query
    if (rateLimitRemaining === 0) {
      this._timeout = rateLimitReset;
    }
  }

  async _fetchJson(endpoint: string): Promise<any> {
    await this._wait();
    const res = await this._fetch(endpoint);
    failIfMissing(res);
    failForNotOk(res);
    this._checkRateLimit(res);
    return await res.json();
  }

  async guilds(): Promise<$ReadOnlyArray<Model.Guild>> {
    const guilds = await this._fetchJson("/users/@me/guilds");
    return guilds.map((x) => ({
      id: x.id,
      name: x.name,
      permissions: x.permissions,
    }));
  }

  async emojis(guild: Model.Snowflake): Promise<$ReadOnlyArray<Model.Emoji>> {
    const emojis = await this._fetchJson(`/guilds/${guild}/emojis`);
    return emojis.map((x) => ({
      id: x.id,
      name: x.name,
    }));
  }

  async channels(
    guild: Model.Snowflake
  ): Promise<$ReadOnlyArray<Model.Channel>> {
    const channels = await this._fetchJson(`/guilds/${guild}/channels`);
    return channels.map((x) => ({
      id: x.id,
      parentId: x.parent_id,
      name: x.name,
      type: Model.channelTypeFromId(x.type),
      nsfw: x.nsfw,
    }));
  }

  async roles(guild: Model.Snowflake): Promise<$ReadOnlyArray<Model.Role>> {
    const roles = await this._fetchJson(`/guilds/${guild}/roles`);
    return roles.map((x) => ({
      id: x.id,
      name: x.name,
    }));
  }

  async members(
    guild: Model.Snowflake
  ): Promise<$ReadOnlyArray<Model.GuildMember>> {
    const limit = 1000;
    let doneLoading = false;
    let allMembers = [];
    let after = "0";
    while (!doneLoading) {
      const newMembers: Array<Model.GuildMember> = await this._fetchJson(
        `/guilds/${guild}/members?after=${after}&limit=${limit}`
      );
      if (newMembers.length < limit) {
        doneLoading = true;
      } else {
        after = newMembers[newMembers.length - 1].user.id;
      }
      allMembers = [...allMembers, ...newMembers];
    }
    return allMembers.map((x) => ({
      user: {
        id: x.user.id,
        username: x.user.username,
        discriminator: x.user.discriminator,
        bot: x.user.bot || x.user.system || false,
      },
      nick: x.nick || null,
      roles: x.roles,
    }));
  }

  async messages(
    channel: Model.Snowflake,
    after: Model.Snowflake,
    limit: number
  ): Promise<$ReadOnlyArray<Model.Message>> {
    const messages = await this._fetchJson(
      `/channels/${channel}/messages?after=${after}&limit=${limit}`
    );
    return messages.map((x) => ({
      id: x.id,
      channelId: channel,
      authorId: x.author.id,
      timestampMs: Date.parse(x.timestamp),
      content: x.content,
      reactionEmoji: (x.reactions || []).map((r) => r.emoji),
      nonUserAuthor: x.webhook_id != null || false,
      mentions: (x.mentions || []).map((user) => ({
        userId: user.id,
        count: Array.from(x.content.matchAll(`<@!?${user.id}>`)).length,
      })),
    }));
  }

  async reactions(
    channel: Model.Snowflake,
    message: Model.Snowflake,
    emoji: Model.Emoji
  ): Promise<$ReadOnlyArray<Model.Reaction>> {
    let doneLoading = false;
    let allReactingUsers: Array<Model.User> = [];
    let after = "0";
    const limit = 100;
    const emojiRef = Model.emojiToRef(emoji);
    while (!doneLoading) {
      const newReactingUsers: Array<Model.User> = await this._fetchJson(
        `/channels/${channel}/messages/${message}/reactions/${emojiRef}?after=${after}&limit=${limit}`
      );
      if (newReactingUsers.length < limit) {
        doneLoading = true;
      } else {
        after = newReactingUsers[newReactingUsers.length - 1].id;
      }
      allReactingUsers = [...allReactingUsers, ...newReactingUsers];
    }

    return allReactingUsers.map((x) => ({
      channelId: channel,
      messageId: message,
      emoji,
      authorId: x.id,
    }));
  }
}

function failIfMissing(response: Response) {
  if (response.status === 404) {
    throw new Error(`404 Not Found on: ${response.url}; maybe bad serverUrl?`);
  }
  if (response.status === 403) {
    throw new Error(`403 Forbidden: bad API username or key?\n${response.url}`);
  }
  if (response.status === 410) {
    throw new Error(`410 Gone`);
  }
}

function failForNotOk(response: Response) {
  if (!response.ok) {
    throw new Error(`not OK status ${response.status} on ${response.url}`);
  }
}
