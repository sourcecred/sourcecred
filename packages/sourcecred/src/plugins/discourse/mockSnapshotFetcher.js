// @flow

import deepFreeze from "deep-freeze";
import {createHash} from "crypto";
import path from "path";
import fs from "fs-extra";
import {Fetcher, type DiscourseFetchOptions} from "./fetch";

export const options: DiscourseFetchOptions = deepFreeze({
  serverUrl: "https://sourcecred-test.discourse.group",
});

async function snapshotFetch(url: string | Request | URL): Promise<Response> {
  const snapshotDir = "src/plugins/discourse/snapshots";
  const shasum = createHash("sha256");
  shasum.update(Buffer.from(((url: any): string), "utf8"));
  const filename = shasum.digest("hex");
  const file = path.join(snapshotDir, filename);
  if (await fs.exists(file)) {
    const contents = await fs.readFile(file);
    return new Response(contents, {status: 200});
  } else {
    throw new Error(`couldn't load snapshot for ${file}`);
  }
}
export const snapshotFetcher = (): Fetcher =>
  new Fetcher(options, snapshotFetch, 0);
