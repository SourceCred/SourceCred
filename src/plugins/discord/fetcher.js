// @flow

import fetch from "isomorphic-fetch";
import * as Model from "./models";

export interface DiscordApi {
  guilds(): Promise<$ReadOnlyArray<Model.Guild>>;
}

const fetcherDefaults: FetcherOptions = {
  apiUrl: "https://discordapp.com/api",
  token: null,
  fetch,
};

type FetcherOptions = {|
  +apiUrl: string,
  +fetch: typeof fetch,
  +token: ?Model.BotToken,
|};

export class Fetcher implements DiscordApi {
  +_options: FetcherOptions;

  constructor(opts?: $Shape<FetcherOptions>) {
    this._options = {...fetcherDefaults, ...opts};
    if (!this._options.token) {
      throw new Error("A BotToken is required");
    }
  }

  _fetch(endpoint: string): Promise<Response> {
    const {apiUrl, token} = this._options;
    if (!token) {
      throw new Error("A BotToken is required");
    }
    const requestOptions = {
      method: "GET",
      headers: {
        Accept: "application/json",
        Authorization: `Bot ${token}`,
      },
    };
    return this._options.fetch(`${apiUrl}${endpoint}`, requestOptions);
  }

  async guilds(): Promise<$ReadOnlyArray<Model.Guild>> {
    const res = await this._fetch("/users/@me/guilds");
    failIfMissing(res);
    failForNotOk(res);
    return (await res.json()).map((g) => ({
      id: g.id,
      name: g.name,
      permissions: g.permissions,
    }));
  }
}

function failIfMissing(response: Response) {
  if (response.status === 404) {
    throw new Error(`404 Not Found on: ${response.url}; maybe bad serverUrl?`);
  }
  if (response.status === 403) {
    throw new Error(`403 Forbidden: bad API username or key?`);
  }
  if (response.status === 410) {
    throw new Error(`410 Gone`);
  }
}

function failForNotOk(response: Response) {
  if (!response.ok) {
    throw new Error(`not OK status ${response.status} on ${response.url}`);
  }
}
