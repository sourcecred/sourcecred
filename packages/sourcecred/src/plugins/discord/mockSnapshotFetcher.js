// @flow

import path from "path";
import fs from "fs-extra";
import {Fetcher} from "./fetcher";
import {createHash} from "crypto";

async function snapshotFetch(url: string | Request | URL): Promise<Response> {
  const snapshotDir = "src/plugins/discord/snapshots";
  const shasum = createHash("sha256");
  shasum.update(Buffer.from(((url: any): string), "utf8"));
  const filename = shasum.digest("hex");
  const file = path.join(snapshotDir, filename);
  if (await fs.exists(file)) {
    const contents = await fs.readFile(file);
    return new Response(contents, {status: 200});
  } else {
    throw new Error(`couldn't load snapshot for ${file} (${String(url)})`);
  }
}

export const snapshotFetcher = (): Fetcher =>
  new Fetcher({
    fetch: snapshotFetch,
    token: "mock-token",
  });
