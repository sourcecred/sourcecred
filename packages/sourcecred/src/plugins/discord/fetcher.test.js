// @flow

import {snapshotFetcher} from "./mockSnapshotFetcher";
import {Fetcher} from "./fetcher";

describe("plugins/discord/fetcher", () => {
  describe("snapshot testing", () => {
    const guildId = "678348980639498428";
    const channelId = "678394406507905129";
    const messageId = "678394436757094410";
    const emoji = {id: "678399364418502669", name: "sourcecred"};

    it("loads guilds", async () => {
      expect(await snapshotFetcher().guilds()).toMatchSnapshot();
    });
    it("loads channels", async () => {
      expect(await snapshotFetcher().channels(guildId)).toMatchSnapshot();
    });
    it("loads emojis", async () => {
      expect(await snapshotFetcher().emojis(guildId)).toMatchSnapshot();
    });
    it("loads roles", async () => {
      expect(await snapshotFetcher().roles(guildId)).toMatchSnapshot();
    });
    it("loads members", async () => {
      expect(await snapshotFetcher().members(guildId)).toMatchSnapshot();
    });
    it("loads messages", async () => {
      expect(
        await snapshotFetcher().messages(channelId, "", 10)
      ).toMatchSnapshot();
      expect(
        await snapshotFetcher().messages(channelId, "678394455849566208", 10)
      ).toMatchSnapshot();
    });
    it("loads reactions", async () => {
      expect(
        await snapshotFetcher().reactions(channelId, messageId, emoji)
      ).toMatchSnapshot();
    });
  });
  describe("_checkRateLimit", () => {
    it("waits when no calls remain", () => {
      const spy = jest.spyOn(global, "setTimeout");
      const currentTime = Math.round(Date.now()) / 1000;
      const mockRes = new Response(null, {
        headers: {
          "x-ratelimit-remaining": "0",
          "x-ratelimit-reset": `${currentTime + 2}`,
        },
      });
      const mockToken = "XYZ";
      const fetcher = new Fetcher({token: mockToken});
      fetcher._checkRateLimit(mockRes);
      //  testing async is flaky, so we just ensure a timeout is actually set
      fetcher._wait();
      expect(setTimeout).toBeCalled();
      expect(spy.mock.calls[0][1]).toBeGreaterThan(0);
      spy.mockRestore();
    });
  });
});
