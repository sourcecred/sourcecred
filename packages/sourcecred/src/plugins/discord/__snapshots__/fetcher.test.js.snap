// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/discord/fetcher snapshot testing loads channels 1`] = `
Array [
  Object {
    "id": "678348980756938813",
    "name": "Text Channels",
    "nsfw": false,
    "parentId": null,
    "type": "GUILD_CATEGORY",
  },
  Object {
    "id": "678348980798881844",
    "name": "Voice Channels",
    "nsfw": false,
    "parentId": null,
    "type": "GUILD_CATEGORY",
  },
  Object {
    "id": "678348980849213472",
    "name": "general",
    "nsfw": false,
    "parentId": "678348980756938813",
    "type": "GUILD_TEXT",
  },
  Object {
    "id": "678348980878573661",
    "name": "General",
    "nsfw": false,
    "parentId": "678348980798881844",
    "type": "GUILD_VOICE",
  },
  Object {
    "id": "678394406507905129",
    "name": "pagination",
    "nsfw": false,
    "parentId": "678348980756938813",
    "type": "GUILD_TEXT",
  },
  Object {
    "id": "678696874869522446",
    "name": "notmymessage",
    "nsfw": false,
    "parentId": "678348980756938813",
    "type": "GUILD_TEXT",
  },
  Object {
    "id": "830088231538130964",
    "name": "rules",
    "nsfw": false,
    "parentId": null,
    "type": "GUILD_TEXT",
  },
  Object {
    "id": "830088231538130965",
    "name": "moderator-only",
    "nsfw": false,
    "parentId": null,
    "type": "GUILD_TEXT",
  },
  Object {
    "id": "830089044695187586",
    "name": "test stage",
    "nsfw": false,
    "parentId": "678348980798881844",
    "type": "UNKNOWN",
  },
  Object {
    "id": "837851969795129344",
    "name": "test-nsfw",
    "nsfw": true,
    "parentId": "678348980756938813",
    "type": "GUILD_TEXT",
  },
]
`;

exports[`plugins/discord/fetcher snapshot testing loads emojis 1`] = `
Array [
  Object {
    "id": "678399364418502669",
    "name": "sourcecred",
  },
]
`;

exports[`plugins/discord/fetcher snapshot testing loads guilds 1`] = `
Array [
  Object {
    "id": "678348980639498428",
    "name": "SourceCred Test Server",
    "permissions": 33620992,
  },
]
`;

exports[`plugins/discord/fetcher snapshot testing loads members 1`] = `
Array [
  Object {
    "nick": null,
    "roles": Array [
      "678350026946117694",
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "5887",
      "id": "143776454050709505",
      "username": "Beanow",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678350026946117694",
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "2630",
      "id": "258786206387666944",
      "username": "Thena",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "8826",
      "id": "318552163649454081",
      "username": "hz",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678350026946117694",
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "3370",
      "id": "420341518948237331",
      "username": "dandelion",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678350026946117694",
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "8658",
      "id": "432981598858903585",
      "username": "wchargin",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678350026946117694",
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "2386",
      "id": "439050857921904640",
      "username": "Brian Litwin",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678350026946117694",
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "1337",
      "id": "489154203714060300",
      "username": "topocount",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678350026946117694",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "3969",
      "id": "579830927287386126",
      "username": "sandpiper",
    },
  },
  Object {
    "nick": null,
    "roles": Array [],
    "user": Object {
      "bot": false,
      "discriminator": "7902",
      "id": "664562775770791966",
      "username": "Beez",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678359229433905152",
    ],
    "user": Object {
      "bot": true,
      "discriminator": "1705",
      "id": "678351352770068560",
      "username": "CredBot-Beanow",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "728740243926286386",
    ],
    "user": Object {
      "bot": true,
      "discriminator": "3557",
      "id": "728738433199112280",
      "username": "sts-credbot",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "767909888226492448",
    ],
    "user": Object {
      "bot": true,
      "discriminator": "9034",
      "id": "758370311774404618",
      "username": "topocount's server bot",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "761706197606334535",
    ],
    "user": Object {
      "bot": true,
      "discriminator": "6627",
      "id": "761705425870782516",
      "username": "SourceCred Test",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "772890579632652339",
    ],
    "user": Object {
      "bot": true,
      "discriminator": "3070",
      "id": "772886503830323210",
      "username": "sourcecred-test-thena",
    },
  },
  Object {
    "nick": null,
    "roles": Array [
      "678349848684003359",
    ],
    "user": Object {
      "bot": false,
      "discriminator": "4121",
      "id": "773595284894253136",
      "username": "amrro",
    },
  },
]
`;

exports[`plugins/discord/fetcher snapshot testing loads messages 1`] = `
Array [
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "20",
    "id": "678394498098659349",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812252307,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "19",
    "id": "678394493124476939",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812251121,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "18",
    "id": "678394489391415348",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812250231,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "17",
    "id": "678394486644146187",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812249576,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "16",
    "id": "678394484291141700",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812249015,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "15",
    "id": "678394480184918016",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [
      Object {
        "id": null,
        "name": "😆",
      },
    ],
    "timestampMs": 1581812248036,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "14",
    "id": "678394477106167818",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812247302,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "13",
    "id": "678394473428025354",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812246425,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "12",
    "id": "678394469048909841",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812245381,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "11",
    "id": "678394468373626890",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812245220,
  },
]
`;

exports[`plugins/discord/fetcher snapshot testing loads messages 2`] = `
Array [
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "9",
    "id": "678394451462193154",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812241188,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "8",
    "id": "678394448497082388",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812240481,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "7",
    "id": "678394445351092275",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812239731,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "6",
    "id": "678394442301833247",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812239004,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "5",
    "id": "678394436757094410",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [
      Object {
        "id": "678399364418502669",
        "name": "sourcecred",
      },
    ],
    "timestampMs": 1581812237682,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "4",
    "id": "678394433233747978",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812236842,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "3",
    "id": "678394431149178940",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812236345,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "2",
    "id": "678394428569813013",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [],
    "timestampMs": 1581812235730,
  },
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "content": "1",
    "id": "678394426153893948",
    "mentions": Array [],
    "nonUserAuthor": false,
    "reactionEmoji": Array [
      Object {
        "id": null,
        "name": "👍",
      },
    ],
    "timestampMs": 1581812235154,
  },
]
`;

exports[`plugins/discord/fetcher snapshot testing loads reactions 1`] = `
Array [
  Object {
    "authorId": "143776454050709505",
    "channelId": "678394406507905129",
    "emoji": Object {
      "id": "678399364418502669",
      "name": "sourcecred",
    },
    "messageId": "678394436757094410",
  },
  Object {
    "authorId": "439050857921904640",
    "channelId": "678394406507905129",
    "emoji": Object {
      "id": "678399364418502669",
      "name": "sourcecred",
    },
    "messageId": "678394436757094410",
  },
]
`;

exports[`plugins/discord/fetcher snapshot testing loads roles 1`] = `
Array [
  Object {
    "id": "678348980639498428",
    "name": "@everyone",
  },
  Object {
    "id": "678349848684003359",
    "name": "allow-adding-data",
  },
  Object {
    "id": "678350026946117694",
    "name": "admins",
  },
  Object {
    "id": "678359229433905152",
    "name": "CredBot-Beanow",
  },
  Object {
    "id": "728740243926286386",
    "name": "sts-credbot",
  },
  Object {
    "id": "761706197606334535",
    "name": "SourceCred Test",
  },
  Object {
    "id": "767909888226492448",
    "name": "SC server bot",
  },
  Object {
    "id": "772890579632652339",
    "name": "sourcecred-test",
  },
]
`;
